File: uva/101-the-blocks-problem/main.c
#include <stdio.h>
#include <stdlib.h>

#define Si(a) scanf("%d", &a)
#define Ss(a) scanf("%s", a)

#define MAX 25

typedef struct block {
  struct block *above;
  struct block *below;
  int table_loc;
  int value;
} block;

block *table[MAX];

void put_back(block *a) {
  block *b;
  while(a->above != NULL) {
    b = a->above;
    a->above = NULL;
    b->below = NULL;
    b->table_loc = b->value;
    table[b->value] = b;
    a = b;
  }
}

block *top(struct block *a) {
  while(a->above != NULL) a = a->above;
  return a;
}

void stack(block *a, block *b) {
  block *top_b = top(b);
  block *a_below = a->below;
  int table_loc = b->table_loc;

  /* Remove stack a from its current position. */
  if(a_below != NULL) a_below->above = NULL;
  else table[a->value] = NULL;

  /* Stack stack a onto the top of stack b. */
  top_b->above = a;
  a->below = top_b;

  /* Update stack reference for block a and blocks above a. */
  while(a != NULL) {
    a->table_loc = table_loc;
    a = a->above;
  }
}

void print_table(int n) {
  block *a;
  int i;
  for(i = 0; i < n; i++) {
    printf("%d:", i);
    for(a = table[i]; a != NULL; a = a->above)  {
      printf(" %d", a->value);
    }
    printf("\n");
  }
}

int main() {
  int n, i, v;
  char verb[4], prep[4];
  block *blocks[MAX];
  block *a, *b;

  for(i = 0; i < MAX; i++)
    table[i] = blocks[i] = NULL;

  /* Read number of blocks `n`. */
  Si(n);

  /* Initialize world with `n` blocks. */
  for(i = 0; i < n; i++) {
    table[i] = blocks[i] = (block *) malloc(sizeof(block));
    blocks[i]->above = NULL;
    blocks[i]->below = NULL;
    blocks[i]->table_loc = i;
    blocks[i]->value = i;
  }

  while(1) {

    /* Read the verb. */
    Ss(verb);

    /* Exit when verb is quit. */
    if(verb[0] == 'q') break;

    /* Read the object. */
    Si(v);
    a = blocks[v];
    
    /* Read the preposition. */
    Ss(prep);

    /* Read the complement. */
    Si(v);
    b = blocks[v];

    /* Ignore action if it is illegal. */
    if(a->table_loc == b->table_loc) continue;

    /* When verb is `move`, put back that are above the object `a`. */
    if(verb[0] == 'm') put_back(a);

    /* When preposition is `onto`, put back the blocks that are above the complement `b`. */
    if(prep[1] == 'n') put_back(b);

    /* Pile object `a` over complement `b`. */
    stack(a, b);
  }

  print_table(n);

  for(i = 0; i < n; i++)
    free(blocks[i]);

  return 0;
}

File: uva/101-the-blocks-problem/python/main.py
#!/usr/bin/env python

'''
move a onto b

Example:

  2   5
  0   4
  1   3
-----------
0 1 2 3 4 5

move 4 onto 1

stack of 4: 3
stack of 1: 1

put back 5
put back 2, 0

      4
0 1 2 3   5
-----------
0 1 2 3 4 5

pile 4 over 1

  4
0 1 2 3   5
-----------
0 1 2 3 4 5
'''

'''
move a over b

Example:

  2   5
  0   4
  1   3
-----------
0 1 2 3 4 5

move 4 over 1

stack of 4: 3
stack of 1: 1

put back 5

  2
  0   4
  1   3   5
-----------
0 1 2 3 4 5

pile 4 over 1

  4
  2
  0
  1   3   5
-----------
0 1 2 3 4 5

'''

'''
pile a onto b

Example:

  2   5
  0   4
  1   3
-----------
0 1 2 3 4 5

pile 4 onto 1

stack of 4: 3
stack of 1: 1

put back 2, 0

      5
      4
0 1 2 3
-----------
0 1 2 3 4 5

pile 4 over 1

  5
  4
0 1 2 3
-----------
0 1 2 3 4 5
'''

'''
pile a over b

Example:

  2   5
  0   4
  1   3
-----------
0 1 2 3 4 5

pile 4 over 1

stack of 4: 3
stack of 1: 1

top of stack 1: 2

set next 3 to null
set next 2 to 4
set prev 4 to 2

  2   5
  0   4
  1   3
-----------
0 1 2 3 4 5
'''

import re

B = []

def stack_of(a):
    c = B[a]
    while True:
        n = c['next']
        if n == None: break
        c = n
    return c['label']

def put_back(a):
    c = B[a]
    while True:
        n = c['next']
        if n == None: break
        n['home'] = True
        c = n

def pile_over(a, b):
    aa = B[a]
    sb = B[stack_of(b)]
    # unstack aa
    if aa['prev']:
        aa['prev']['next'] = None
    # stack aa over sb
    sb['next'] = aa
    aa['prev'] = sb

def print_stack(c):
    while True:
        print c['label'],
        n = c['next']
        if n == None: break
        c = n

def print_table():
    for b in B:
        print "%d:" % b['label'],
        if b['prev'] == None:
            print_stack(b)
        print

def main():
    with open ('/dev/stdin', 'rt') as f:
        n = int(f.readline().rstrip())
#        print 'n =', n
        for i in range(0, n):
            B.append({'label': i, 'next': None, 'prev': None})
        while True:
            line = f.readline().rstrip()
#            print 'line =', line
            if line.strip() == 'quit':
                break
            v, a, p, b = re.split(r' +', line)
            a = int(a)
            b = int(b)
            # is this a valid command?
            if stack_of(a) == stack_of(b):
                # ignore the command
                continue
            if v == 'move':
                put_back(a)
            if p == 'onto':
                put_back(b)
            pile_over(a, b)
#            print_table()
        print_table()

main()
